sys/usb/usbus/msc/msc.c

```diff
@@ -0,0 +1,453 @@
+        if (msc->state == DATA_TRANSFER_IN) {
+            msc->state = GEN_CSW;
+        }
+    }
+}
+
+static void _handle_rx_event(event_t *ev)
+{
+    usbus_msc_device_t *msc = container_of(ev, usbus_msc_device_t, rx_event);
+
+    _read_xfer(msc);
+}
+
+static usbus_msc_device_t *_get_msc_handler(usbus_t *usbus)
+{
+    for (usbus_interface_t *iface = usbus->iface; iface; iface = iface->next) {
+        if (iface->class == USB_CLASS_MASS_STORAGE) {
+            return (usbus_msc_device_t *)iface->handler;
+        }
+    }
+    return NULL;
+}
+
+static unsigned _get_lun(usbus_msc_device_t *msc)
+{
+    unsigned count = 0;
+
+    /* Count only registered MTD devices as USB LUN, (using usbus_msc_add_lun)
+     * not every MTD devices available on board */
+    for (unsigned i = 0; i < USBUS_MSC_EXPORTED_DEV; i++) {
+        if (msc->lun_dev[i].mtd != NULL) {
+            count++;
+        }
+    }
+    return count;
+}
+
+int usbus_msc_add_lun(usbus_t *usbus, mtd_dev_t *dev)
+{
+    uint32_t block_size;
+    /* Loop through all interfaces to find the MSC one */
+    usbus_msc_device_t *msc = _get_msc_handler(usbus);
+
+    if (!dev) {
+        return -ENODEV;
+    }
+
+    /* Check if MTD isn't already registered */
+    for (unsigned i = 0; i < USBUS_MSC_EXPORTED_DEV; i++) {
+        if (dev == msc->lun_dev[i].mtd) {
+            return -EBUSY;
+        }
+    }
+    /* Store new MTD device in first slot available
+       Also re alloc internal buffer if needed */
+    for (unsigned i = 0; i < USBUS_MSC_EXPORTED_DEV; i++) {
+        if (msc->lun_dev[i].mtd == NULL) {
+            int ret = mtd_init(dev);
+            if (ret != 0) {
+                DEBUG_PUTS("[msc]: cannot initialize MTD device");
+                return -ENODEV;
+            }
+            block_size = dev->page_size * dev->pages_per_sector;
+            /* Ensure MTD device can be exported */
+            if (block_size * dev->sector_count < 512) {
+                DEBUG("page size:%ld pages per sector:%ld sector_count:%ld\n", dev->page_size,
+                      dev->pages_per_sector, dev->sector_count);
+                return -ENOMEM;
+            }
+            if (msc->buffer != NULL) {
+                /* If new registered MTD device need more memory than the
+                   previous, realloc a new buffer */
+                if (block_size > msc->buffer_size) {
+                    msc->buffer = realloc(msc->buffer, block_size);
+                }
+            } else {
+                /* No buffer allocated yet, so allocate one */
+                msc->buffer = malloc(block_size);
+            }
+
+            if (!msc->buffer) {
+                DEBUG_PUTS("[msc]: Failed to realloc new buffer\n");
+                return -ENOMEM;
+            }
+            msc->buffer_size = block_size;
+            msc->lun_dev[i].mtd = dev;
+            return 0;
+        }
+    }
+
+    /* No more slots available */
+    return -EAGAIN;
```

Comment:
```suggestion
    return -ENFILE;
```
might be a better fit